### DISCUSSION TITLE
Super Slow under Load with file responce

### DISCUSSION THREAD

**Discussion by BluchipStudio** [fastapi/fastapi#8418] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get("/download/{channel_id}/{base_url:path}", response_class=FileResponse)
async def download_segments(channel_id: str, base_url: str):
    
    URL_PHARSE  = urlparse(base_url)
    FILE_NAME   = os.path.basename(URL_PHARSE.path)
    
    if "/audio/" in base_url:
        SEGMENT_TYPE = "audio"
    else:
        SEGMENT_TYPE = "video"

    SEGMENT_ID      = FILE_NAME.replace(".m4s","")
    CONTENT_TYPE    = SEGMENT_TYPE + "/mp4"

    FILE_LOCATION = await GET_CACHE_PATH("segments" ,True, channel_id) + SEGMENT_ID + "_" + SEGMENT_TYPE + ".m4s"

  return FileResponse(FILE_LOCATION, media_type=CONTENT_TYPE)
```

### Description

I am trying to create a dash streaming server using fast api and uvicorn everything works absolutly perfect and a custom dash manifest is built from my main endpoint and then the second endpoint in the example grabs and provides the segments back to the client

my issue comes when there are more than 4 clients trying to view media as soon as 4 or more people try connecting then the streams start to stutter and pause as soon as it drops back down to 3 users or below the streams work perfect again 

the way thesystem works is i have all the segments in ramdisk storage and when a user tries to play it in a media player the download section of the code is called and the file is provided back to the user to me it seems that once 4 or more people connect the operation of returning the files slows down drastically 

### Operating System

Linux

### Operating System Details

0S : UBUNTU 18.04
RAM : 128GB
STORAGE : 500GB SSD
BANDWIDTH : 1GBPS DEDICATED

### FastAPI Version

Latest

### Python Version

3.10

### Additional Context

i have the same kind of system up and running on another server using nginx and php and it works good no issues even with 30/40 people connected but the upkeep and amount of code needed is like 100x that of python that is why i was looking at switching to fast api

**Comment by jgould22** [comment]
I can't think of anything immediately obvious but here are some things I would check, in no particular order
  
Look at the file response object and see if it is behaving in a way you do not expect.
https://github.com/encode/starlette/blob/master/starlette/responses.py#L279

If you dont notice anything in the above try opening the file yourself and  use just a regular response object 

You should probably check to see you are not saturating your disk access in some way with something like `iotop` or exhausting your ram and causing things to page to disk with something like top/htop

Also how are you deploying this? How many workers are you running? If you are just running the single python process is working hard or not ?

**Comment by BluchipStudio** [comment]
> I can't think of anything immediately obvious but here are some things I would check, in no particular order
> 
> Look at the file response object and see if it is behaving in a way you do not expect. https://github.com/encode/starlette/blob/master/starlette/responses.py#L279
> 
> If you dont notice anything in the above try opening the file yourself and use just a regular response object
> 
> You should probably check to see you are not saturating your disk access in some way with something like `iotop` or exhausting your ram and causing things to page to disk with something like top/htop
> 
> Also how are you deploying this? How many workers are you running? If you are just running the single python process is working hard or not ?

I am running this locally now using uvicorn in the python code and starting it with python3 main.py as the app will only be used by devices connected to the same network

i am using a tempfs ram disk so there are no read/right io operations to the sdd and thus saves the storage the tempfs ramdisk is 40gb big and 1 stream never uses more than 90mb of storage at any given time

i have checked htop and cpu/ram usage when the app is running stays nice and low like you would expect just reading and outputting files from a ramdisk

**Comment by jgould22** [comment]
Probably I would try with just a regular response, and set everything manually and see if that side steps the issue.

Also make sure you don't have any blocking code anywhere that might be blocking the event loop in some way.

**Comment by BluchipStudio** [comment]
> Probably I would try with just a regular response, and set everything manually and see if that side steps the issue.
> 
> Also make sure you don't have any blocking code anywhere that might be blocking the event loop in some way.

i will give it a try with a normal responce now as for blocking code i dont think there is any as everything runs fine for 1-3 open streams anything over is what kills it thats why i was thinking it had to be something todo with the responce times that got worse as more was opened

**Comment by jgould22** [comment]
Well if you have blocking calls happening that effectively makes the event loop pause and wait for those calls to complete and it will not make any progress on other tasks/requests while it is blocked.  More requests would equal more blocking which equals more pausing.

Is you program doing any non-async file or network io or using a library that might spend a long time between calls to functions with `await`

edit: You could find out pretty quick by changing your `async def download_segments(channel_id: str, base_url: str):` to `def download_segments(channel_id: str, base_url: str):` (without the async) which will cause fast api to run the request on a threadpool

**Comment by davismartens** [comment]
Have you tried using `gunicorn` instead of `uvicorn`? Launching more workers might help when you have multiple clients access your method.
